### ISSUE TITLE
Formatter destroys Config while formatting

### ISSUE THREAD

**Issue by cdietrich** [eclipse/xtext-core#1190]
given the grammar

```
grammar org.xtext.example.mydsl.MyDsl with org.eclipse.xtext.common.Terminals

generate myDsl "http://www.xtext.org/example/mydsl/MyDsl"

Model:
	"spec" name=ID "{"
	"}";

```

and formatter (old)

```
	@Inject
	private MyDslGrammarAccess grammarAccess;

	@Override
	protected void configureFormatting(FormattingConfig c) {
		// It's usually a good idea to activate the following three statements.
		// They will add and preserve newlines around comments
		c.setLinewrap(0, 1, 2).before(grammarAccess.getSL_COMMENTRule());
		c.setLinewrap(0, 1, 2).before(grammarAccess.getML_COMMENTRule());
		c.setLinewrap(0, 1, 1).after(grammarAccess.getML_COMMENTRule());
		c.setLinewrap(2).between(grammarAccess.getModelAccess().getLeftCurlyBracketKeyword_2(),
				grammarAccess.getSL_COMMENTRule());
	}
```

the following test will show that for formatter formats differently the first and second time

```
import com.google.inject.Inject
import org.eclipse.xtext.formatting.INodeModelFormatter
import org.eclipse.xtext.formatting.INodeModelFormatter.IFormattedRegion
import org.eclipse.xtext.nodemodel.ICompositeNode
import org.eclipse.xtext.resource.XtextResource
import org.eclipse.xtext.testing.InjectWith
import org.eclipse.xtext.testing.extensions.InjectionExtension
import org.eclipse.xtext.testing.util.ParseHelper
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.^extension.ExtendWith
import org.xtext.example.mydsl.myDsl.Model

@ExtendWith(InjectionExtension)
@InjectWith(MyDslInjectorProvider)
class MyDslParsingTest {
	@Inject
	ParseHelper<Model> parseHelper
	
	@Inject
	INodeModelFormatter f
	
	@Test
	def void loadModel() {
		val result = parseHelper.parse('''

    spec specWithComments {//demo
    // test single signal wins over group

}

		''')
		Assertions.assertNotNull(result)
		val errors = result.eResource.errors
		Assertions.assertTrue(errors.isEmpty, '''Unexpected errors: ?errors.join(", ")?''')
		
		var ICompositeNode rootNode = (result.eResource as XtextResource).getParseResult().getRootNode();
		var IFormattedRegion iformattedRegion = f.format(rootNode, rootNode.getTotalOffset(),
                rootNode.getTotalLength());
        var String actual = iformattedRegion.getFormattedText();
        println(actual) 
        System.err.println("=================================")
		rootNode = (result.eResource as XtextResource).getParseResult().getRootNode();
		iformattedRegion = f.format(rootNode, rootNode.getTotalOffset(),
                rootNode.getTotalLength());
        actual = iformattedRegion.getFormattedText();
        println(actual) 
	}
}
```

this is caused by `org.eclipse.xtext.formatting.impl.FormattingConfigBasedStream.collectLocatorsForComments(Collection<ElementLocator>, EObject, EObject)`
which removes stuff from the list `leadingElementLocators = cfg.getLocatorsForCommentTokensBefore(right);`

=> either there should be a copy of the list created here or `org.eclipse.xtext.formatting.impl.AbstractFormattingConfig.getLocatorsForCommentTokensBefore(EObject)` should return a copied list

**Comment by cdietrich** [comment]
@meysholdt  do you know why this is done this way.
and should the list be copied in the AbstractFormattingConfig or in collectLocatorsForComments

**Comment by cdietrich** [comment]
here is the workflow to easily reproduce

```
module org.xtext.example.mydsl.GenerateMyDsl

import org.eclipse.xtext.xtext.generator.*
import org.eclipse.xtext.xtext.generator.model.project.*

var rootPath = ".."

Workflow {
	
	component = XtextGenerator {
		configuration = {
			project = StandardProjectConfig {
				baseName = "org.xtext.example.mydsl"
				rootPath = rootPath
				runtimeTest = {
					enabled = true
				}
				eclipsePlugin = {
					enabled = true
				}
				eclipsePluginTest = {
					enabled = true
				}
				createEclipseMetaData = true
			}
			code = {
				encoding = "US-ASCII"
				lineDelimiter = "\n"
				fileHeader = "/*\n * generated by Xtext \${version}\n */"
			}
		}
		language = StandardLanguage {
			name = "org.xtext.example.mydsl.MyDsl"
			fileExtensions = "mydsl"

			serializer = {
				generateStub = false
			}
			validator = {
				// composedCheck = "org.eclipse.xtext.validation.NamesAreUniqueValidator"
				// Generates checks for @Deprecated grammar annotations, an IssueProvider and a corresponding PropertyPage
				generateDeprecationValidation = true
			}
			junitSupport = {
				junitVersion = "5"
			}
			fragment = org.eclipse.xtext.generator.adapter.FragmentAdapter {
				fragment = org.eclipse.xtext.generator.formatting.FormatterFragment {}
			}
		}
	}
}

```

**Comment by cdietrich** [comment]
Fixed in 2.19
